src/protocol/attribution/mod.rs

```diff
@@ -231,12 +239,66 @@ where
     Ok(())
 }
 
+async fn compute_helper_bits_gf2<C, S>(
+    ctx: C,
+    sorted_match_keys: &[Vec<S>],
+) -> Result<Vec<S>, Error>
+where
+    C: Context,
+    S: LinearSecretSharing<Gf2> + BasicProtocols<C, Gf2>,
+{
+    let narrowed_ctx = ctx
+        .narrow(&Step::ComputeHelperBits)
+        .set_total_records(sorted_match_keys.len() - 1);
+
+    try_join_all(
+        sorted_match_keys
+            .iter()
+            .zip(sorted_match_keys.iter().skip(1))
```

Comment:
I found a cool technique for this where we want to have a "sliding window".

```
sorted_match_keys.windows(2).enumerate().map(|(i, rows)| {
    ...
    async move { bitwise_equal_gf2(c, record_id, rows[0], rows[1]).await }
})
```

Reply:
I think that you need to do this:

```rust
sorted_match_keys/*.iter() ?*/.windows(2).enumerate().map(|(i, [first, second])| {
  async move { bitwise_equal_gf2(c, RecordId::from(i), second, first).await }
```

The drawback here is that you need a slice.  This won't work otherwise, though with the double-iteration requirement, that's maybe inevitable.

A specific adapter might be even better.

```rust
let mut it = sorted_match_keys.into_iter(); // See note.
let mut prev = it.next().ok_or(Error::NoDataWtf)?;
it.map(|v| { let r = (prev, v); prev = v; r }).enumerate(|i, (first, second)| {
  // ... as above ...
});
```

Note that I'm using `into_iter()` here and assuming that the values are `Copy`.  That's not ideal (or even true), but it simplifies things.  Otherwise, you need to clone every item in the list.

You could maybe do this with references and avoid the clone, but it might be a little fiddly.

Reply:
https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=fad373f3c12798257189c36d21eb9733

Reply:
I pulled the PR and tried the following. Seems to work okay :)

```
    try_join_all(sorted_match_keys.windows(2).enumerate().map(|(i, rows)| {
        let c = narrowed_ctx.clone();
        let record_id = RecordId::from(i);
        async move { bitwise_equal_gf2(c, record_id, &rows[0], &rows[1]).await }
    }))
    .await
```

There is `array_windows` that yields an iterator like in Martin's suggestion, but I think we are okay with just `windows()`.

Reply:
That doesn't work generically though, sadly.  Try this:

https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=99a63b7f87733b461183b8454bfffc83

Reply:
Homework for someone who wants a challenge: implement this for pairs that aren't adjacent so that we can to the GraphSC thing over a small window.  That won't help for capping (which has to traverse the entire thing, such that a slice is necessary), but it might be OK for some applications.

Reply:
That is so cool @taikiy! Thanks for this code snippet, I've added it in!